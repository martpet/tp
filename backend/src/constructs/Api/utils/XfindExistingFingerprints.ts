import {
  BatchGetCommand,
  DynamoDBClient,
  DynamoDBDocument,
  photosTableOptions,
  unmarshall,
} from 'lambda-layer';

const ddbClient = new DynamoDBClient({});
const ddbDocClient = DynamoDBDocument.from(ddbClient);

export const findExistingFingerprints = async (fingerprints: string[]) => {
  const { tableName, partitionKey } = photosTableOptions;
  const result: string[] = [];

  async function batchGet(Keys: Record<string, any>[]) {
    const batchGetCommand = new BatchGetCommand({
      RequestItems: { [tableName]: { Keys } },
    });
    const { Responses, UnprocessedKeys } = await ddbDocClient.send(batchGetCommand);

    if (Responses) {
      Responses[tableName].forEach(({ fingerprint }) => result.push(fingerprint));
    }
    if (UnprocessedKeys) {
      const nextKeys = UnprocessedKeys[tableName]?.Keys?.map((key) => unmarshall(key));
      if (nextKeys) {
        await batchGet(nextKeys);
      }
    }
  }

  const batchSize = 100;
  const requests = [];

  for (let i = 0; i < fingerprints.length; i += batchSize) {
    const chunk = fingerprints.slice(i, i + batchSize);
    const keys = chunk.map((fingerprint) => ({ [partitionKey.name]: fingerprint }));
    requests.push(batchGet(keys));
  }

  await Promise.all(requests);
  return result;
};
