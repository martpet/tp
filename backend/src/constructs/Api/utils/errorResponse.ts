import { APIGatewayProxyResult } from 'aws-lambda';
import { getReasonPhrase, StatusCodes } from 'http-status-codes';
import { SetRequired } from 'type-fest';

type ErrorResponseOptions = WithExposedError | WithoutExposedError;

type BaseOptions = {
  description?: string;
  error?: unknown;
  statusCode?: StatusCodes;
};

type WithExposedError = SetRequired<BaseOptions, 'error'> & {
  exposeError: true;
};

type WithoutExposedError = BaseOptions & {
  exposeError?: never | false;
};

export const errorResponse = (
  traceId: string,
  {
    statusCode = StatusCodes.INTERNAL_SERVER_ERROR,
    description,
    error,
    exposeError,
  }: ErrorResponseOptions = {}
): APIGatewayProxyResult => {
  const { envName } = globalLambdaProps;

  const errorObj: Record<string, unknown> = {
    statusCode,
    message: getReasonPhrase(statusCode),
    traceId,
  };

  if (description) {
    errorObj.description = description;
  }

  if (error) {
    console.error(`[${traceId}]`, error);

    if (exposeError ?? envName !== 'production') {
      errorObj.error = String(error);
    }
  }

  return {
    statusCode,
    body: JSON.stringify({ error: errorObj }),
  };
};
