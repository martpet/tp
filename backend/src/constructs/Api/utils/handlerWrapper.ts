import { APIGatewayProxyHandlerV2 } from 'aws-lambda';

export const handlerWrapper =
  (func: APIGatewayProxyHandlerV2) =>
  (...params: Parameters<APIGatewayProxyHandlerV2>) =>
    func(...params);
