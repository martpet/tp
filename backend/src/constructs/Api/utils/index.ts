export * from './cookieName';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './lambdaEdgeErrorResponse';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
