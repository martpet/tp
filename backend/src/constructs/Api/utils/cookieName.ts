import { AllApiRoutesCookies } from '~/constructs/Api/types';

export const cookieName = (name: keyof AllApiRoutesCookies) => name;
