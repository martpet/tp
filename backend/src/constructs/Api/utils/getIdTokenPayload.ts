import camelcaseKeys from 'camelcase-keys';

import { IdTokenPayload } from '~/constructs/Api/types';

export const getIdTokenPayload = (idToken: string): IdTokenPayload => {
  const payload = JSON.parse(Buffer.from(idToken.split('.')[1], 'base64').toString());

  return camelcaseKeys(payload);
};
