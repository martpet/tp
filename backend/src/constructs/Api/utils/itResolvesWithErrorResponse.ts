import { APIGatewayProxyHandlerV2 } from 'aws-lambda';

import { errorResponse } from '~/constructs/Api/utils';

export const itResolvesWithErrorResponse = (
  handler: APIGatewayProxyHandlerV2,
  args: Parameters<APIGatewayProxyHandlerV2>
) => {
  it('calls "errorResponse" with correct args', async () => {
    await handler(...args);
    expect(vi.mocked(errorResponse).mock.calls).toMatchSnapshot();
  });

  it('resolves with a correct value', () => {
    return expect(handler(...args)).resolves.toMatchSnapshot();
  });
};
