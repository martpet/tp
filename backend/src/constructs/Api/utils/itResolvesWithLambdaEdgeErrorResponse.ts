import { LambdaEdgeViewerRequestHandler } from '~/constructs/Api/types';
import { lambdaEdgeErrorResponse } from '~/constructs/Api/utils';

export const itResolvesWithLambdaEdgeErrorResponse = (
  handler: LambdaEdgeViewerRequestHandler,
  args: Parameters<LambdaEdgeViewerRequestHandler>
) => {
  it('calls "lambdaEdgeErrorResponse" with correct args', async () => {
    await handler(...args);
    expect(vi.mocked(lambdaEdgeErrorResponse).mock.calls).toMatchSnapshot();
  });

  it('resolves with a correct value', () => {
    return expect(handler(...args)).resolves.toMatchSnapshot();
  });
};
