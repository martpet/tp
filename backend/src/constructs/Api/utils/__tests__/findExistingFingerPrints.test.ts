import { BatchGetCommand, DynamoDBDocumentClient } from '@aws-sdk/lib-dynamodb';
import { mockClient } from 'aws-sdk-client-mock';

import { itResolves, itSendsAwsCommand } from '~/constructs/Api/utils';
import { photosTableOptions } from '~/consts';

import { findExistingFingerprints } from '../XfindExistingFingerprints';

vi.mock('@aws-sdk/util-dynamodb');

const ddbMock = mockClient(DynamoDBDocumentClient);

const args = [['fingerprint1']] as Parameters<typeof findExistingFingerprints>;

beforeEach(() => {
  ddbMock.reset();

  ddbMock.on(BatchGetCommand).resolves({
    Responses: {
      [photosTableOptions.tableName]: [
        { fingerprint: 'existingFingerprint1' },
        { fingerprint: 'existingFingerprint1' },
      ],
    },
  });
});

describe('findExistingFingerprints', () => {
  itSendsAwsCommand(BatchGetCommand, ddbMock, findExistingFingerprints, args);
  itResolves(findExistingFingerprints, args);

  describe('when called with 101 items', async () => {
    const argsClone = structuredClone(args);
    argsClone[0] = Array(101);

    it('sends "BatchGetCommand" 2 times', async () => {
      await findExistingFingerprints(...argsClone);
      expect(ddbMock.commandCalls(BatchGetCommand).length).toBe(2);
    });
  });

  describe('when "BatchGetCommand" response is missing "Responses"', () => {
    beforeEach(() => {
      ddbMock.on(BatchGetCommand).resolves({});
    });
    itResolves(findExistingFingerprints, args);
  });

  describe('when "BatchGetCommand" response has "UnprocessedKeys"', () => {
    beforeEach(() => {
      ddbMock
        .on(BatchGetCommand)
        .resolvesOnce({
          Responses: {
            [photosTableOptions.tableName]: [{ fingerprint: 'existingFingerprint3' }],
          },
          UnprocessedKeys: {
            [photosTableOptions.tableName]: { Keys: [{ dummyUnprocessedKey: '' }] },
          },
        })
        .resolvesOnce({
          Responses: {
            [photosTableOptions.tableName]: [{ fingerprint: 'existingFingerprint4' }],
          },
        });
    });

    it('sends "BatchGetCommand" a second time with correct args', async () => {
      await findExistingFingerprints(...args);
      expect(ddbMock.commandCalls(BatchGetCommand)[1].args[0].input).toMatchSnapshot();
    });

    itResolves(findExistingFingerprints, args);
  });
});
