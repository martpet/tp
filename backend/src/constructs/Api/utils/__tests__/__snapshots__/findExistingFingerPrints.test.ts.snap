// Vitest Snapshot v1

exports[`findExistingFingerprints > resolves with a correct value 1`] = `
[
  "existingFingerprint1",
  "existingFingerprint1",
]
`;

exports[`findExistingFingerprints > sends "BatchGetCommand" from "DynamoDBDocumentClient" with correct args 1`] = `
{
  "RequestItems": {
    "TP-photos": {
      "Keys": [
        {
          "fingerprint": "fingerprint1",
        },
      ],
    },
  },
}
`;

exports[`findExistingFingerprints > when "BatchGetCommand" response has "UnprocessedKeys" > resolves with a correct value 1`] = `
[
  "existingFingerprint3",
  "existingFingerprint4",
]
`;

exports[`findExistingFingerprints > when "BatchGetCommand" response has "UnprocessedKeys" > sends "BatchGetCommand" a second time with correct args 1`] = `
{
  "RequestItems": {
    "TP-photos": {
      "Keys": [
        "dummyUnmarshalledItems",
      ],
    },
  },
}
`;

exports[`findExistingFingerprints > when "BatchGetCommand" response is missing "Responses" > resolves with a correct value 1`] = `[]`;
